diff --git a/manual/conf.py b/manual/conf.py
index 8ff15ae0..a1aced1b 100644
--- a/manual/conf.py
+++ b/manual/conf.py
@@ -20,7 +20,7 @@
 # import sys
 # sys.path.insert(0, os.path.abspath('.'))
 
-import os, sys, git, re, pathlib
+import os, sys, re, pathlib
 
 # -- CAF-specific variables ---------------------------------------------------
 
@@ -39,19 +39,7 @@ with open(os.path.join(root_dir, "libcaf_core/caf/config.hpp")) as f:
     patch = raw_version % 100
     version = '{}.{}.{}'.format(major, minor, patch)
 
-# We're building a stable release if the last commit message is
-# "Change version to <version>".
-repo = git.Repo(root_dir)
-is_stable = repo.head.commit.message.startswith("Change version to " + version)
-
-# Generate the full version, including alpha/beta/rc tags. For stable releases,
-# this is always the same as the CAF version.
-if is_stable:
-    release = version
-else:
-    last_commit_full = str(repo.head.commit)
-    last_commit = last_commit_full[:7]
-    release = version + "+exp.sha." + last_commit
+release = %%VERSION%%
 
 # -- Enable Sphinx to find the literal includes -------------------------------
 
