diff --git a/auxil/zeekctl/CMakeLists.txt b/auxil/zeekctl/CMakeLists.txt
index a2029f5..6e07954 100644
--- a/auxil/zeekctl/CMakeLists.txt
+++ b/auxil/zeekctl/CMakeLists.txt
@@ -266,9 +266,6 @@ InstallPackageConfigFile(
     ${ETC}
     node.cfg)

-# Install legacy BroControl wrappers and links.
-InstallSymlink("${CMAKE_INSTALL_PREFIX}/bin/zeek-wrapper" "${CMAKE_INSTALL_PREFIX}/bin/broctl")
-
 if ( NOT BINARY_PACKAGING_MODE )
   # Need to remove pre-existing broctl dir from previous installs.
   set(_broctl_lib_dst ${LIBDIR}/broctl)
@@ -283,9 +280,6 @@ if ( NOT BINARY_PACKAGING_MODE )
   ")
 endif ()

-InstallSymlink("${PY_MOD_INSTALL_DIR}/zeekctl" "${LIBDIR}/broctl")
-install(DIRECTORY BroControl DESTINATION ${PY_MOD_INSTALL_DIR}/zeekctl)
-
 ########################################################################
 ## Packaging Setup


diff --git a/auxil/broker/include/broker/bro.hh b/auxil/broker/include/broker/bro.hh
deleted file mode 100644
index 83e5620..0000000
--- a/auxil/broker/include/broker/bro.hh
+++ /dev/null
@@ -1,35 +0,0 @@
-#pragma once
-
-#pragma message("Warning: bro.hh header is deprecated, use zeek.hh instead")
-
-#include "broker/zeek.hh"
-
-namespace broker {
-namespace bro {
-
-using Message
-      [[deprecated("use version from zeek.hh and zeek namespace instead")]]
-      = broker::zeek::Message;
-
-using Event
-      [[deprecated("use version from zeek.hh and zeek namespace instead")]]
-      = broker::zeek::Event;
-
-using Batch
-      [[deprecated("use version from zeek.hh and zeek namespace instead")]]
-      = broker::zeek::Batch;
-
-using LogCreate
-      [[deprecated("use version from zeek.hh and zeek namespace instead")]]
-      = broker::zeek::LogCreate;
-
-using LogWrite
-      [[deprecated("use version from zeek.hh and zeek namespace instead")]]
-      = broker::zeek::LogWrite;
-
-using IdentifierUpdate
-      [[deprecated("use version from zeek.hh and zeek namespace instead")]]
-      = broker::zeek::IdentifierUpdate;
-
-} // namespace broker
-} // namespace bro

diff --git a/auxil/broker/bindings/python/broker/__init__.py b/auxil/broker/bindings/python/broker/__init__.py
index 95d6128..63f1834 100644
--- a/auxil/broker/bindings/python/broker/__init__.py
+++ b/auxil/broker/bindings/python/broker/__init__.py
@@ -427,7 +427,6 @@ class Message:
         assert False and "method not overridden"

 from . import zeek
-from . import bro

 class Data(_broker.Data):
     def __init__(self, x = None):
