--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -595,7 +595,7 @@
 endif ()
 
 install(DIRECTORY DESTINATION ${ZEEK_ETC_INSTALL_DIR})
-install(DIRECTORY DESTINATION ${ZEEK_STATE_DIR})
+#install(DIRECTORY DESTINATION ${ZEEK_STATE_DIR})
 install(DIRECTORY DESTINATION ${ZEEK_SPOOL_DIR})
 install(DIRECTORY DESTINATION ${ZEEK_LOG_DIR})
 
@@ -1293,7 +1293,7 @@
     configure_file(${CMAKE_CURRENT_SOURCE_DIR}/cmake_templates/zkg-config.in
                    ${CMAKE_CURRENT_BINARY_DIR}/zkg-config @ONLY)
 
-    install(DIRECTORY DESTINATION var/lib/zkg)
+    #install(DIRECTORY DESTINATION var/lib/zkg)
     install(FILES ${CMAKE_CURRENT_BINARY_DIR}/zkg-config DESTINATION ${ZEEK_ZKG_CONFIG_DIR}
             RENAME config)
 endif ()
