diff --git a/auxil/zeek-aux/rst/rst.c b/auxil/zeek-aux/rst/rst.c
index dbf31b2..cb95a26 100644
--- a/auxil/zeek-aux/rst/rst.c
+++ b/auxil/zeek-aux/rst/rst.c
@@ -113,7 +113,7 @@ in_cksum(register u_short *addr, register int len)

 // - adapted from tcpdump
 // Returns the ones-complement checksum of a chunk of b short-aligned bytes.
-int ones_complement_checksum(const void *p, int b, uint32_t sum)
+int __attribute__ ((noinline)) ones_complement_checksum(const void *p, int b, uint32_t sum)
 {
 	const uint16_t *sp = (uint16_t *) p;	// better be aligned!

