--- a/Makefile
+++ b/Makefile
@@ -22,17 +22,12 @@
 
 NL3_EXISTS := $(shell pkg-config --exists libnl-route-3.0 && echo yes)
 
-COMMON_FLAGS += -O2 -c \
-	-D_GNU_SOURCE -D_FILE_OFFSET_BITS=64 \
-	-fPIE \
-	-Wformat -Wformat-security -Wno-format-nonliteral \
-	-Wall -Wextra -Werror \
-	-Ikafel/include
+COMMON_FLAGS += -c -D_GNU_SOURCE -Ikafel/include
 
 CXXFLAGS += $(USER_DEFINES) $(COMMON_FLAGS) $(PROTOBUF_CFLAGS) \
-	-std=c++20 -fno-exceptions -Wno-unused -Wno-unused-parameter
+	-std=c++20 -fno-exceptions
 
-LDFLAGS += -pie -Wl,-z,noexecstack -lpthread $(PROTOBUF_LIBS)
+LDFLAGS += -lpthread $(PROTOBUF_LIBS)
 
 ifeq ($(NL3_EXISTS), yes)
 	CXXFLAGS += $(shell pkg-config --cflags libnl-route-3.0)
@@ -86,7 +81,7 @@
 # Standard Object Compilation
 # The | $(SRCS_PB_H) ensures headers exist before we try to compile .cc files
 %.o: %.cc | $(SRCS_PB_H)
-	$(CXX) $(CXXFLAGS) $< -o $@
+	$(CXX) $(CPPFLAGS) $(CXXFLAGS) $< -o $@
 
 # Protobuf Generation
 # We only define the recipe for the .cc file to prevent race conditions.
