diff --git a/pkt_netflow.c b/pkt_netflow.c
index d474e88..f323fa3 100644
--- a/pkt_netflow.c
+++ b/pkt_netflow.c
@@ -5129,7 +5129,7 @@ static int __init pkt_netflow_init(void)
 #endif
 						      );
 #else /* 2.6.25 */
-	netflow_sysctl_header = register_sysctl_paths(netflow_sysctl_path, netflow_sysctl_table);
+	netflow_sysctl_header = register_sysctl("net/netflow", netflow_sysctl_table);
 #endif
 	if (!netflow_sysctl_header) {
 		printk(KERN_ERR "netflow: can't register to sysctl\n");
